waku/node/peer_manager/peer_manager.nim

```diff
@@ -283,108 +289,101 @@ proc canBeConnected*(pm: PeerManager,
   let now = Moment.init(getTime().toUnix, Second)
   let lastFailed = pm.peerStore[LastFailedConnBook][peerId]
   let backoff = calculateBackoff(pm.initialBackoffInSec, pm.backoffFactor, failedAttempts)
-  if now >= (lastFailed + backoff):
-    return true
-  return false
+ 
+  return now >= (lastFailed + backoff)
 
 ##################
 # Initialisation #
 ##################
 
 proc getPeerIp(pm: PeerManager, peerId: PeerId): Option[string] =
-  if pm.switch.connManager.getConnections().hasKey(peerId):
-    let conns = pm.switch.connManager.getConnections().getOrDefault(peerId)
-    if conns.len != 0:
-      let observedAddr = conns[0].connection.observedAddr
-      let ip = observedAddr.get.getHostname()
-      if observedAddr.isSome:
-        # TODO: think if circuit relay ips should be handled differently
-        let ip = observedAddr.get.getHostname()
-        return some(ip)
-  return none(string)
+  if not pm.switch.connManager.getConnections().hasKey(peerId):
+    return none(string)
+
+  let conns = pm.switch.connManager.getConnections().getOrDefault(peerId)
+  if conns.len == 0:
+    return none(string)
+
+  let obAddr = conns[0].connection.observedAddr.valueOr:
+    return none(string)
+
+  # TODO: think if circuit relay ips should be handled differently
+  
+  return some(obAddr.getHostname())
 
 # called when a connection i) is created or ii) is closed
 proc onConnEvent(pm: PeerManager, peerId: PeerID, event: ConnEvent) {.async.} =
   case event.kind
   of ConnEventKind.Connected:
-    let direction = if event.incoming: Inbound else: Outbound
+    #let direction = if event.incoming: Inbound else: Outbound
     discard
   of ConnEventKind.Disconnected:
     discard
 
+proc onPeerMetadata(pm: PeerManager, peerId: PeerId) {.async.} =
+  # To prevent metadata protocol from breaking prev nodes, by now we only
+  # disconnect if the clusterid is specified.
+  if pm.wakuMetadata.clusterId == 0:
+    return
+
+  let res = catch: await pm.switch.dial(peerId, WakuMetadataCodec)
+
+  let reason = 
+    if (let conn = res; conn.isOk()):
+      if (let metadata = (await pm.wakuMetadata.request(conn.get())); metadata.isOk()):
+        if (let clusterId = metadata.get().clusterId; clusterId.isSome()):
+          if pm.wakuMetadata.clusterId == clusterId.get():
+            if metadata.get().shards.anyIt(pm.wakuMetadata.shards.contains(it)):
+              return
+            else: "no shards in common"
+          else: "different clusterId reported: " & $pm.wakuMetadata.clusterId & " vs " & $clusterId
+        else: "empty clusterId reported"
+      else: "failed waku metadata codec request"
+    else: "waku metadata codec not supported"
```

Comment:
This snippet is great indeed! The only thing is that we may bring a little more information because we are missing a couple of result error messages. The following is aimed to enhance that a little :)

```suggestion
  var reason: string
  if res.isErr():
    reason = "waku metadata codec not supported: " & res.error[].msg
  else:
    let metadata = await pm.wakuMetadata.request(res.get())
    if metadata.isErr():
      reason = "failed waku metadata codec request: " & $metadata.error
    else:
      if (let clusterId = metadata.get().clusterId; clusterId.isSome()):
        if pm.wakuMetadata.clusterId == clusterId.get():
          if metadata.get().shards.anyIt(pm.wakuMetadata.shards.contains(it)):
            return
          else: reason = "no shards in common"
        else: reason = "different clusterId: " & $pm.wakuMetadata.clusterId & " vs " & $clusterId
      else: reason = "empty clusterId reported"
```

Reply:
Yes indeed. I already changed it because I wanted more details. commit incoming